cumulus/parachains/common/src/storage_weight_reclaim.rs

```diff
@@ -0,0 +1,497 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Mechanism to reclaim PoV weight after an extrinsic has been applied.
+
+use codec::{Decode, Encode};
+use cumulus_primitives_core::Weight;
+use cumulus_primitives_proof_size_hostfunction::storage_proof_size::storage_proof_size;
+use frame_support::{
+	dispatch::{DispatchInfo, PostDispatchInfo},
+	weights::WeightMeter,
+};
+use frame_system::Config;
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, PostDispatchInfoOf, SignedExtension},
+	transaction_validity::TransactionValidityError,
+	DispatchResult,
+};
+use sp_std::marker::PhantomData;
+
+const LOG_TARGET: &'static str = "runtime::storage_reclaim";
+
+/// Indicates that proof recording was disabled on the node side.
+const PROOF_RECORDING_DISABLED: u64 = u64::MAX;
+
+/// StorageWeightReclaimer is a mechanism for manually reclaiming storage weight.
+///
+/// It internally keeps track of the proof size and storage weight at initialization time. At
+/// reclaim  it computes the real consumed storage weight and refunds excess weight.
+///
+///
+/// # Example
+///
+/// ```ignore
+/// 	use parachains_common::storage_weight_reclaim::StorageWeightReclaimer;
+///
+/// 	let mut remaining_weight_meter = WeightMeter::with_limit(Weight::from_parts(10, 10));
+/// 	let mut reclaim_helper = StorageWeightReclaimer::start(&remaining_weight_meter);
+/// 	remaining_weight_meter.try_consume(get_weight_for_work()).is_ok() {
+/// 		do_work();
+/// 		if let Some(relaimed) = reclaim_helper.reclaim_with_meter(&mut remaining_weight_meter) {
+/// 			log::info!("Reclaimed {} weight", reclaimed_weight);
+/// 		}
+/// 	}
+/// ```
+pub struct StorageWeightReclaimer {
+	previous_remaining_proof_weight: u64,
+	previous_reported_proof_size: Option<u64>,
+}
+
+impl StorageWeightReclaimer {
+	/// Creates a new `StorageWeightReclaimer` instance and initializes it with the current storage
+	/// weight and reported proof size from the node.
+	pub fn start(weight_meter: &WeightMeter) -> StorageWeightReclaimer {
+		let previous_remaining_proof_weight = weight_meter.remaining().proof_size();
+		let previous_reported_proof_size = get_proof_size();
+		Self { previous_remaining_proof_weight, previous_reported_proof_size }
+	}
+
+	/// Check the consumed storage weight and calculate the consumed excess weight.
+	fn reclaim(&mut self, remaining_weight: Weight) -> Option<Weight> {
```

Comment:
(Opinionated nit) Personally, I would find it easier to follow if the following variable names were used:
- `initial_proof_size` -> `previous_storage_proof_size`
- `used_storage_proof` -> `reported_used_size`
- `used_weight` -> `expected_used_size`